### ISSUE TITLE
Error reported from BIND error during parallel transfer should be more verbose

### ISSUE THREAD

**Issue by trel** [irods/irods#2803]
From https://groups.google.com/forum/#!topic/irod-chat/ylqMd3pYsTw:

> we think the error message is due to the checksum being the last thing that failed, but triggered by the lower level problem of not having enough ports open (a BIND error when trying to find an open port for the large file transfer).  We'll work on making the error messages better so clients and servers can figure out what is happening.

A possible additional solution may include a retry with exponential backoff.

**Comment by trel** [comment]
The following three files will generate a loop that puts and gets large files in parallel.

With JOBS=30, the `3-runloop.sh` script usually fails within a couple minutes on a robust machine.

To test:

1-makefiles.sh
```bash
FILESIZE_IN_MB=400
mkdir -p files.dir
cd files.dir
for i in `seq -w 256`
do
    dd if=/dev/zero of=bigfile.$i bs=1M count=$FILESIZE_IN_MB 2> /dev/null
    echo $i >> bigfile.$i
done
```

2-generatescripts.sh
```bash
RESOURCE=demoResc
SCRIPTPATH=$( cd $(dirname $0) ; pwd -P )
FILESDIR=$SCRIPTPATH/files.dir
BENCHMARKSDIR=$SCRIPTPATH/benchmarks.dir
mkdir -p $BENCHMARKSDIR
cd $FILESDIR
for i in `ls`
do
    echo iput -R ${RESOURCE} -K -f $FILESDIR/$i benchmarks.dir/
done > $BENCHMARKSDIR/iput-all
for i in `ls`
do
    echo iget -R ${RESOURCE} -K -f benchmarks.dir/$i
done > $BENCHMARKSDIR/iget-all
```

3-runloop.sh
```bash
JOBS=30
date
irm -rf benchmarks.dir
imkdir -p benchmarks.dir
COUNT=0
time while [ "$?" == "0" ]; do
    COUNT=$((COUNT+1))
    echo "Starting loop $COUNT"
    echo " put"
    cat benchmarks.dir/iput-all | parallel -j $JOBS
    if [ "$?" != "0" ]; then
        exit 1
    fi
    echo " get"
    cat benchmarks.dir/iget-all | parallel -j $JOBS
done
```

**Comment by kript** [comment]
In your tests, are you using a single server for the iCAT, iRES and database, or are they separate? We are seeing issue on 3.3.1, but only when we get about 50-60 concurrent connections, which may be explained by the 4 iRES servers and separate iCAT, Database (and federated iCAT on top), but it would be good to have an idea of the number of concurrent queries we could tell our users to expect to work (while reminding them of the -K flag to iget..:-) )

**Comment by trel** [comment]
The set of tests above were all on a single machine.

**Comment by keithj** [comment]
I've run some similar tests on 4.1.3 with increasing numbers of clients from 32 to 256 (both the iRODS server and ICAT on the same host). I set the server config to allow a range of 10,000 ports.

I `iput` and `iget` 256 128Mb files with increasing numbers of concurrent clients `n`. The number of errors encountered is below for a typical run:

|     n | 32 | 48 |  64 | 80 |  96 | 112 | 128 | 144 | 160 | 176 | 192 | 208 | 224 | 240 | 256 |
|-------|----|----|-----|----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|
| iget |  0 |  0 | 191 |  0 | 226 | 227 | 255 | 256 | 256 | 251 | 253 | 252 | 255 | 252 | 256 |
| iput |  0 |  0 |   0 |  0 |   0 |   1 |   1 |   1 |   1 |   1 |   1 |   0 |   1 |   2 |   1 |


If I add a 2 second delay between starting each client:

|     n | 32 | 48 | 64 | 80 | 96 | 112 | 128 | 144 | 160 | 176 | 192 | 208 | 224 | 240 | 256 |
------|----|----|----|----|----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|
| iget |  0 |  0 |  0 | 17 |  0 |   0 |   0 |   0 |   0 |   0 |   0 |   0 |   0 |   0 |   0 |
| iput |  0 |  0 |  0 |  0 |  0 |   0 |   0 |   0 |   0 |   0 |   0 |   0 |   0 |   0 |   0 |

**Comment by keithj** [comment]
I've realised that I made a stupid mistake running the scripts! I gave parallel an argument that I should not have done and was triggering a lot of fails myself! I'd like to retract those results and get back to you with an accurate number. I apologise for causing head-scratching over there.

The number of fails is much less than I've reported, as you suspected during the con call. I'll post up the script as agreed, but I'll do re-run first. In the meantime I'll remove the table to above to avoid misleading anyone who reads this later.

**Comment by keithj** [comment]
The corrected script is below. (I had given `parallel` a `--halt now,fail=1` argument so that it killed all remaining jobs after the first failure.)

I'm still seeing a lot of failures at 64 clients and above, but I can now see that they are due to client connection failures under load. That is not the type of failure I'm interested in and is confounding measurements.

Scripts adapted from the Google groups post:

````
#!/bin/bash

mkdir random_files.dir

cd random_files.dir

for i in `seq -w 256`
do
    dd if=/dev/urandom of=random_file.128M.$i bs=1M count=128
done

md5sum random_file.128M.* > ../random_files.md5
````

````
#!/bin/bash

RESOURCE=demoResc
N=256 # num files

mkdir -p benchmarks.dir
imkdir -p benchmarks.coll

cd random_files.dir

for i in random_file.128M.*
do
    echo iput -R ${RESOURCE} -K -f $PWD/$i benchmarks.coll/
done > ../benchmarks.dir/iput-all

for i in random_file.128M.*
do
    echo iget -R ${RESOURCE} -f benchmarks.coll/$i
done > ../benchmarks.dir/iget-all

# Benchmark
cd ../benchmarks.dir

for j in `seq 1 3`
do
    for i in `seq 32 16 $N`
    do
        echo
        echo $i iput jobs:
        date
        cat iput-all | ../parallel --jobs $i --joblog iput.$i.$j.log
        date
        echo $i iget jobs:
        date
        cat iget-all | ../parallel --jobs $i --joblog iget.$i.$j.log
        date
        md5sum random_file.128M.* > /tmp/benchmark.md5
        rm -f random_file.128M.*
        echo Checksum errors: `comm -2 -3 <(sort ../random_files.md5) <(sort /tmp/benchmark.md5) | wc -l`
        sleep 5
    done 2>&1 | tee benchmark.$j.log
done
````

**Comment by kellerb** [comment]
@keithj 

> That is not the type of failure I'm interested in and is confounding measurements.

What type of failure are you investigating?

**Comment by keithj** [comment]
I was looking for the data corruption reported in the original Google groups report above, where `iget` could download a file that did not match the checksum (which would be apparent if you used the -K flag).

I wanted to reproduce here the improvement obtained by increasing the number of ports available.

**Comment by kellerb** [comment]
My understanding of this issue is that all of the reported checksum mismatches were caused by (possibly poorly reported) port exhaustion on the server side. Meaning that I don't expect there to be checksum mismatches seen (in 4.1) when `iput` and/or `iget` complete successfully (return code of `0`).

**Comment by keithj** [comment]
Yes, that's what I wanted to reproduce. I didn't expect any failures. What I am seeing is a lot of unrelated noise due to the server being unable to cope with the number of clients in my test. I didn't realise that the server wouldn't be able to handle the numbers I was using.

I think I just need to moderate the number of clients a bit and then I should see what you're describing (no checksum differences when `iput` and `iget` complete successfully).

**Comment by trel** [comment]
After some more testing today with the scripts from above (https://github.com/irods/irods/issues/2803#issuecomment-122273867)...

We have identified two independent issues:

1) A potential timeout in `select()` while waiting for the client to connect to the portal in `acceptSrvPortal()` which we have solved locally with a retry loop.
https://github.com/irods/irods/blob/4.1.7/iRODS/server/core/src/miscServerFunct.cpp#L272

2) Rarely, two rodsAgent processes bind to an identical port number which causes both processes to fail due to the collision.  A solution would be to retry the entire portal transfer mechanism - however, this is too much for now (affects too much code for a rare case).  We are considering randomizing the port selection mechanism rather than a linear scan for an available port.
https://github.com/irods/irods/blob/4.1.7/iRODS/lib/core/src/sockComm.cpp#L362
